components/hab/src/command/pkg/uninstall.rs

```diff
@@ -0,0 +1,199 @@
+/// 3. We do a BFS on the graph to get the dependencies in order
+/// 4. We check if the specified package has any reverse deps
+///     4a. If there are, we throw an error
+///     4b. If not, we delete the package
+/// 5. For each dependency we check if there are any packages which depend on it
+///     5a. If there are, we skip it
+///     5b. If there are not, we delete it from disk and the graph
+///
+pub fn start(
+    ui: &mut UI,
+    ident: &PackageIdent,
+    fs_root_path: &Path,
+    execution_strategy: ExecutionStrategy,
+    scope: Scope,
+) -> Result<()> {
+    // 1.
+    let pkg_install = PackageInstall::load(ident, Some(fs_root_path))?;
+    let ident = pkg_install.ident();
+    ui.begin(format!("Uninstalling {}", &ident))?;
+
+    // 2.
+    let packages = load_all_packages(&fs_root_path)?;
+    let mut graph = PackageGraph::new();
+    graph.build(&packages)?;
+
+    // 3.
+    let deps = graph.ordered_deps(&ident);
+
+    // 4.
+    match graph.count_rdeps(&ident) {
+        None => {
+            // package not in graph - this shouldn't happen but could be a race condition in Step 2 with another hab uninstall. We can
+            // continue as what we wanted (package to be removed) has already happened. We're going to continue and try and delete down through the
+            // dependency tree anyway
+            ui.warn(format!("Tried to find dependant packages of {} but it wasn't in graph.  Maybe another uninstall command was run at the same time?", &ident))?;
+        }
+        Some(0) => {
+            maybe_delete(ui, &fs_root_path, &pkg_install, &execution_strategy)?;
+            graph.remove(&ident);
+        }
+        Some(c) => {
+            return Err(Error::CannotRemovePackage(ident.clone(), c));
+        }
+    }
+
+    // 5.
+    let mut count = 0;
+    match scope {
+        Scope::Package => {
+            ui.status(Status::Skipping, "dependencies (--no-deps specified)")?;
+        }
+        Scope::PackageAndDependencies => {
+            for p in &deps {
+                match graph.count_rdeps(&p) {
+                    None => {
+                        // package not in graph - this shouldn't happen but could be a race condition in Step 2 with another hab uninstall. We can
+                        // continue as what we wanted (package to be removed) has already happened. We're going to continue and try and delete down through the
+                        // dependency tree anyway
+                        ui.warn(format!("Tried to find dependant packages of {} but it wasn't in graph.  Maybe another uninstall command was run at the same time?", &p))?;
+                    }
+                    Some(0) => {
+                        let install = packages.iter().find(|&i| i.ident() == p).unwrap();
+                        if let Ok(true) =
+                            maybe_delete(ui, &fs_root_path, &install, &execution_strategy)
+                        {
+                        }
```

Comment:
What are these empty braces supposed to do? Are the following two lines supposed to be inside them?

Reply:
It was an unneeded `if let` and `rust fmt` added them.  Changed the code to not use the if let.